Pickpocketing

LethalLeonard opened the pull request:
## What has been done?
Added the pickpocketing part of the Thieving skill. All normal pickpocket-able NPCs are included, but not quest NPCs. 

Edit: Item tables are done to the best accuracy I could. All that may be wrong is the exact amounts that the weighting should be done by.

LethalLeonard commented:
Cheers! Glad it will be useful!

LethalLeonard commented:
Uh, that was a rebase.

DennisSnijder commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocketing/pickpocket.kts:
```diff
@@ -0,0 +1,69 @@
+import gg.rsmod.plugins.content.combat.isAttacking
+import gg.rsmod.plugins.content.combat.isBeingAttacked
+
+//for each npc in the enum
```
The code kinda speaks for itself making some comments redundant.

DennisSnijder commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocketing/pickpocket.kts:
```diff
@@ -0,0 +1,69 @@
+    npcClass.ids.forEach { npcID ->
+
+        //Extra check to ensure that the npc is pickpocket-able
+        if (world.definitions.get(NpcDef::class.java, npcID).options.contains("Pickpocket")) {
```
Rather seeing this like `if(!world.defi......) {continue;}` 
the foreach statements and if statements are already stacking up quite a bit so this would make it a bit cleaner.

DennisSnijder commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocketing/pickpocket.kts:
```diff
@@ -0,0 +1,69 @@
+                        if (player.inventory.hasSpace) {
+                            //checks if the player has the required level for the npc
+                            val thievLvl: Int = player.getSkills().getCurrentLevel(Skills.THIEVING)
+                            if (thievLvl >= npcClass.lvl) {
```
Same for this one, just check if the thieving level is smaller and do a `continue;`

LethalLeonard commented:
All fixed in 6b42613

Tomm0017 commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocketing/pickpocket.kts:
```diff
@@ -0,0 +1,61 @@
+                    player.message("...and you have failed.")
+
+                    //damages player for a value in the npc's damage range
+                    player.hit(npcClass.damage.random(), HitType.HIT)
```
Since you defined the ``HitType`` as a ``HIT``, it would show all damage from this hit as a red hitsplat including 0s. The ``hit`` method does not require you to define a ``HitType``, but does allow it for flexibility purposes.

```suggestion
                    player.hit(npcClass.damage.random())
```

Tomm0017 commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocketing/PickpocketInfo.kt:
```diff
@@ -0,0 +1,461 @@
+ * @ids = an array of NPC ids for them to pickpocket
+ * @exp = the amount of experience given per pickpocket
+ * @lvl = the level requirement to pickpocket that npc
+ * @npcName = the name of the NPC for the chat messages
```
```suggestion
 npc.getDef().name // will allow for npc.name in an upcoming commit
```

Tomm0017 replied:
``npc.name`` is acceptable as of https://github.com/Tomm0017/rsmod/commit/65ee09236b20bf8a2be116d0f4c1b76214fa3f55

LethalLeonard commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocketing/PickpocketInfo.kt:
```diff
@@ -0,0 +1,461 @@
+ * @exp = the amount of experience given per pickpocket
+ * @lvl = the level requirement to pickpocket that npc
+ * @npcName = the name of the NPC for the chat messages
+ * @rewards = a weighted set of possible item rewards
```
I don't understand the benefit of this if you could clarify. If I were to make them an Array, that just adds an extra level of conversion that isn't needed, no?

Tomm0017 replied:
There's several benefits to taking this approach, including:

- Remove boilerplate code (``add`` on every single ``WeightItem``)
- Less error-prone
- If we ever want to change the rewards to __not__ use ``WeightItemSet`` you only have to remove it in a single place instead of every entry

LethalLeonard commented:
Fixed all of those, though, with NPC names, I added that in the kotlin script file, and just set a default for npcName in the enum to make sure I have support for Fremennik citizens and elves keeping the proper name for pickpocketing actions as on OSRS.

LethalLeonard commented:
Fixed an error in pickpocket.kts, corrected filename conventions, and changed the Rarity enums to constants. Should be complete.

Tomm0017 commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocket/PickpocketNpc.kt:
```diff
@@ -0,0 +1,462 @@
+private const val COMMON = 256
+private const val UNCOMMON = 32
+private const val RARE = 8
+private const val VERY_RARE = 1
```
These should be inside the PickpocketNpcs enum class, otherwise they will be globally exposed to other files. Should be defined inside the companion object as they are constants

notmeta commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocket/PickpocketNpc.kt:
```diff
@@ -0,0 +1,462 @@
+ * @damage = damage range when getting stunned
+ * @stunTicks = the amount of time that the npc stuns the player for
+ */
+enum class PickpocketNpcs(val npcIds: IntArray, val experience: Double, val lvl: Int, val npcName: String? = null,
```
The naming should be singular (`PickpocketNpc` vs `PickpocketNpcs`) - see https://stackoverflow.com/a/15756009/5065094

notmeta replied:
While I'm here:
- rename `npcName` just to `name`
- similarly, rename `npcIds` to `ids`
- rename `lvl` to `levelRequirement` (or similar, so that it's clear it's a thieving level requirement)

LethalLeonard replied:
As stated here, @Tomm0017 wanted npcIds to be like this, as it was previously ids. https://github.com/Tomm0017/rsmod/pull/59#discussion_r269592540

notmeta commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocket/pickpocket.kts:
```diff
@@ -0,0 +1,64 @@
+}
+
+suspend fun QueueTask.pickpocket(npcId: Int, npcType: PickpocketNpcs) {
+    val thievLvl: Int = player.getSkills().getCurrentLevel(Skills.THIEVING)
```
`thievLvl` -> `thievingLevel`
Type declaration not necessary

notmeta commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocket/pickpocket.kts:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+suspend fun QueueTask.pickpocket(npcId: Int, npcType: PickpocketNpcs) {
```
`npcType` -> `npc`

notmeta commented on game/plugins/src/main/kotlin/gg/rsmod/plugins/content/skills/thieving/pickpocket/pickpocket.kts:
```diff
@@ -0,0 +1,64 @@
+
+    //determine if the pickpocket was successful or not by "if random number is within success chances"
+    val bonus = if (player.hasEquipped(EquipmentType.GLOVES, Items.GLOVES_OF_SILENCE)) GLOVES_OF_SILENCE_BONUS else 0
+    if (thievLvl.interpolate(55, 95, npcType.lvl, 99, 100 - bonus)) {
```
Personally I would probably separate this logic and have it in a separate method/location as it is pretty important. That way it can be documented better and not just discretely shoved in like this.

LethalLeonard commented:
All done, thank you @notmeta
